lib/agent.js

```diff
@@ -56,8 +56,15 @@ Agent.prototype.close = function(err) {
 };
 
 Agent.prototype._cleanup = function() {
+
+  // Only clean up once if the stream emits both 'end' and 'close'.
+  if (this.closed) return;
```

Comment:
Nice, this will be defensive against 'end' accidentally being emitted twice by the stream too, which the old code didn't do.